src/vsr/journal.zig

```diff
@@ -1356,93 +1355,141 @@ pub fn JournalType(comptime Replica: type, comptime Storage: type) type {
+            if (op_max < replica.op_checkpoint()) return torn_slots[0..0];
 
-            const header_untrusted = &journal.headers_redundant[torn_slot.index];
-            const header = header_ok(replica.cluster, torn_slot, header_untrusted) orelse
-                return null;
-            // The redundant header is valid, also for the correct cluster and not misdirected.
+            // Nothing to truncate - replica could not have prepared anything beyond prepare_max.
+            if (op_max == replica.op_prepare_max()) return torn_slots[0..0];
 
-            if (header.operation == .reserved) {
-                // This is the first log cycle.
+            const op_checkpoint = replica.op_checkpoint();
+            const op_prepare_max = replica.op_prepare_max();
+            const op_prepare_max_slot = journal.slot_for_op(op_prepare_max);
+            const op_checkpoint_slot = journal.slot_for_op(op_checkpoint);
 
-                // TODO Can we be more sure about this? What if op_max is clearly many cycles ahead?
-                // Any previous log cycle is then expected to have a prepare, not a reserved header,
-                // unless the prepare header was lost, in which case this slot may also not be torn.
-            } else {
-                // The redundant header was already written, so the prepare is corrupt, not torn.
-                if (header.op == torn_op) return null;
+            assert(op_max < op_prepare_max);
 
-                assert(header.op < torn_op); // Since torn_op > op_max.
-                // The redundant header is from any previous log cycle.
-            }
+            // Range is constructed such that the op for all *valid* prepares or headers in it
+            // should be less than op_max. If a prepare/header within this range is corrupted, that
+            // makes our op_max uncertain.
+            const op_max_to_op_prepare_max = SlotRange{
+                .head = journal.slot_for_op(op_max + 1),
+                .tail = prepare_max: {
+                    if (op_checkpoint > 0 and op_max == op_checkpoint) {
+                        assert(op_prepare_max_slot.index == op_checkpoint_slot.index);
+                        assert(op_prepare_max > 0);
 
-            const checkpoint_index = journal.slot_for_op(replica.op_checkpoint()).index;
-            const known_range = SlotRange{
-                .head = Slot{ .index = checkpoint_index },
-                .tail = torn_slot,
+                        break :prepare_max journal.slot_for_op(op_prepare_max - 1);
+                    } else {
+                        break :prepare_max op_prepare_max_slot;
+                    }
+                },
             };
-
-            // We must be certain that the torn prepare really was being appended to the WAL.
-            // Return null if any faults do not lie between the checkpoint and the torn prepare,
-            // such as:
+            var torn_slot_count: u8 = 0;
```

Comment:
```suggestion
            var torn_slot_count: u32 = 0;
```

`journal_iops_write_max < maxInt(u8)` and `journal_slot_count < maxInt(u8)`, but that isn't an invariant we depend on elsewhere afaik.

Reply:
Returning bounded array obviates the need for this variable.